File: scripts/node/yali_inspiration.mjs
#!/usr/bin/env node
import http from "node:http";
import https from "node:https";

const DEFAULT_BASE_URL = "https://www.yaliai.com/wp-json/yali/v1";
const HTTP_USER_AGENT = "Mozilla/5.0 (Yali AI Skill Inspiration) AppleWebKit/537.36 Chrome/120 Safari/537.36";
const USAGE = `Usage:
  node scripts/node/yali_inspiration.mjs [--base-url URL] search --query TEXT [--limit N] [--offset N] [--category NAME] [--dry-run]
  node scripts/node/yali_inspiration.mjs [--base-url URL] categories [--dry-run]
  node scripts/node/yali_inspiration.mjs [--base-url URL] case --case-id CASE_ID [--dry-run]
  node scripts/node/yali_inspiration.mjs [--base-url URL] templates [--dry-run]`;

function die(message, code = 2) {
  console.error(JSON.stringify({ ok: false, error: message }));
  process.exit(code);
}

function parseArgs(argv) {
  if (argv.includes("--help") || argv.includes("-h")) {
    console.log(USAGE);
    process.exit(0);
  }
  const global = { baseUrl: DEFAULT_BASE_URL };
  let i = 0;
  while (i < argv.length && argv[i].startsWith("--")) {
    const key = argv[i].slice(2).replace(/-([a-z])/g, (_, c) => c.toUpperCase());
    const value = argv[i + 1];
    if (value === undefined || value.startsWith("--")) die(`Missing value for ${argv[i]}`);
    global[key] = value;
    i += 2;
  }
  const command = argv[i];
  if (!command || !["search", "categories", "case", "templates"].includes(command)) {
    die("Usage: yali_inspiration.mjs [--base-url URL] <search|categories|case|templates> ...");
  }
  const args = {};
  for (i += 1; i < argv.length; i += 1) {
    const item = argv[i];
    if (!item.startsWith("--")) die(`Unexpected argument: ${item}`);
    const key = item.slice(2).replace(/-([a-z])/g, (_, c) => c.toUpperCase());
    if (key === "dryRun") {
      args.dryRun = true;
      continue;
    }
    const value = argv[i + 1];
    if (value === undefined || value.startsWith("--")) die(`Missing value for ${item}`);
    args[key] = value;
    i += 1;
  }
  return { global, command, args };
}

function cleanBaseUrl(value) {
  return String(value || DEFAULT_BASE_URL).trim().replace(/\/+$/, "");
}

function getJson(url) {
  const client = url.startsWith("https:") ? https : http;
  return new Promise((resolve, reject) => {
    const req = client.get(url, { headers: { Accept: "application/json", "User-Agent": HTTP_USER_AGENT } }, (res) => {
      const chunks = [];
      res.on("data", (chunk) => chunks.push(chunk));
      res.on("end", () => {
        const text = Buffer.concat(chunks).toString("utf8");
        if (res.statusCode < 200 || res.statusCode >= 300) {
          reject(new Error(`HTTP ${res.statusCode}: ${text.slice(0, 500)}`));
          return;
        }
        try {
          resolve(JSON.parse(text));
        } catch {
          reject(new Error(`Non-JSON response: ${text.slice(0, 500)}`));
        }
      });
    });
    req.setTimeout(60000, () => req.destroy(new Error(`Timeout calling ${url}`)));
    req.on("error", reject);
  });
}

function normalizeSearch(payload) {
  const items = payload?.items || payload?.response?.items || [];
  const promptFor = (item) => (typeof item.prompt === "string" ? item.prompt.trim() : "");
  return {
    ok: true,
    count: Array.isArray(items) ? items.length : 0,
    items: Array.isArray(items) ? items.map((item) => ({
      case_id: item.case_id,
      slug: item.slug,
      title: item.title,
      category: item.category,
      categories: item.categories,
      keywords: item.keywords,
      prompt: promptFor(item),
      image_url: item.image_url,
      thumb_url: item.thumb_url,
      detail_url: item.detail_url,
      source: item.source,
    })) : [],
  };
}

function normalizeTemplates(payload) {
  const presets = payload?.templatePresets || payload?.response?.templatePresets || {};
  return {
    ok: true,
    count: Object.keys(presets).length,
    defaultTemplateKey: payload?.defaultTemplateKey || payload?.response?.defaultTemplateKey,
    templatePresets: presets,
    capabilities: payload?.capabilities || payload?.response?.capabilities || {},
    raw: payload,
  };
}

async function main() {
  const { global, command, args } = parseArgs(process.argv.slice(2));
  const baseUrl = cleanBaseUrl(global.baseUrl);
  let url;
  if (command === "search") {
    if (!args.query && !args.q) die("search requires --query.");
    url = `${baseUrl}/inspiration/search?${new URLSearchParams({ q: args.query || args.q, limit: args.limit || "5", ...(args.category ? { category: args.category } : {}), ...(args.offset ? { offset: args.offset } : {}) })}`;
    if (args.dryRun) {
      console.log(JSON.stringify({ ok: true, dry_run: true, endpoint: url }, null, 2));
      return;
    }
    console.log(JSON.stringify(normalizeSearch(await getJson(url)), null, 2));
    return;
  }
  if (command === "categories") {
    url = `${baseUrl}/inspiration/categories`;
    if (args.dryRun) {
      console.log(JSON.stringify({ ok: true, dry_run: true, endpoint: url }, null, 2));
      return;
    }
    console.log(JSON.stringify({ ok: true, response: await getJson(url) }, null, 2));
    return;
  }
  if (command === "case") {
    if (!args.caseId) die("case requires --case-id.");
    url = `${baseUrl}/inspiration/cases/${encodeURIComponent(args.caseId)}`;
    if (args.dryRun) {
      console.log(JSON.stringify({ ok: true, dry_run: true, endpoint: url }, null, 2));
      return;
    }
    console.log(JSON.stringify({ ok: true, response: await getJson(url) }, null, 2));
    return;
  }
  if (command === "templates") {
    url = `${baseUrl}/free-image/api/templates`;
    if (args.dryRun) {
      console.log(JSON.stringify({ ok: true, dry_run: true, endpoint: url }, null, 2));
      return;
    }
    console.log(JSON.stringify(normalizeTemplates(await getJson(url)), null, 2));
  }
}

main().catch((error) => die(error.message || String(error), 1));

File: scripts/python/yali_inspiration.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import sys
import urllib.parse
import urllib.request
from typing import Any


DEFAULT_BASE_URL = "https://www.yaliai.com/wp-json/yali/v1"
HTTP_USER_AGENT = "Mozilla/5.0 (Yali AI Skill Inspiration) AppleWebKit/537.36 Chrome/120 Safari/537.36"


def die(message: str, code: int = 2) -> None:
    print(json.dumps({"ok": False, "error": message}, ensure_ascii=False), file=sys.stderr)
    raise SystemExit(code)


def clean_base_url(value: str | None) -> str:
    return (value or DEFAULT_BASE_URL).strip().rstrip("/")


def get_json(url: str) -> dict[str, Any]:
    request = urllib.request.Request(url, headers={"Accept": "application/json", "User-Agent": HTTP_USER_AGENT})
    try:
        with urllib.request.urlopen(request, timeout=60) as response:
            raw = response.read().decode("utf-8")
    except Exception as exc:
        die(f"Inspiration API request failed: {exc}", code=1)
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        die(f"Inspiration API returned non-JSON response: {raw[:500]}", code=1)
    if not isinstance(payload, dict):
        die("Inspiration API returned JSON that is not an object.", code=1)
    return payload


def response_items(payload: dict[str, Any]) -> list[Any]:
    response = payload.get("response") if isinstance(payload.get("response"), dict) else {}
    items = response.get("items", payload.get("items", []))
    return items if isinstance(items, list) else []


def normalize_search(payload: dict[str, Any]) -> dict[str, Any]:
    items = response_items(payload)
    normalized = []
    for item in items:
        if not isinstance(item, dict):
            continue
        prompt = item.get("prompt").strip() if isinstance(item.get("prompt"), str) else ""
        normalized.append(
            {
                "case_id": item.get("case_id"),
                "slug": item.get("slug"),
                "title": item.get("title"),
                "category": item.get("category"),
                "categories": item.get("categories"),
                "keywords": item.get("keywords"),
                "prompt": prompt,
                "image_url": item.get("image_url"),
                "thumb_url": item.get("thumb_url"),
                "detail_url": item.get("detail_url"),
                "source": item.get("source"),
            }
        )
    return {"ok": True, "count": len(normalized), "items": normalized}


def normalize_templates(payload: dict[str, Any]) -> dict[str, Any]:
    response = payload.get("response") if isinstance(payload.get("response"), dict) else {}
    presets = response.get("templatePresets", payload.get("templatePresets", {}))
    if not isinstance(presets, dict):
        presets = {}
    return {
        "ok": True,
        "count": len(presets),
        "defaultTemplateKey": response.get("defaultTemplateKey", payload.get("defaultTemplateKey")),
        "templatePresets": presets,
        "capabilities": response.get("capabilities", payload.get("capabilities", {})),
        "raw": payload,
    }


def print_or_dry_run(url: str, dry_run: bool, normalizer=None) -> None:
    if dry_run:
        print(json.dumps({"ok": True, "dry_run": True, "endpoint": url}, ensure_ascii=False, indent=2))
        return
    payload = get_json(url)
    print(json.dumps(normalizer(payload) if normalizer else {"ok": True, "response": payload}, ensure_ascii=False, indent=2))


def main() -> int:
    parser = argparse.ArgumentParser(description="Query Yali inspiration endpoints with normalized JSON output.")
    parser.add_argument("--base-url", default=DEFAULT_BASE_URL)
    subparsers = parser.add_subparsers(dest="command", required=True)

    search_parser = subparsers.add_parser("search")
    search_parser.add_argument("--query", "--q", dest="query", required=True)
    search_parser.add_argument("--limit", default="5")
    search_parser.add_argument("--offset", default=None)
    search_parser.add_argument("--category", default=None)
    search_parser.add_argument("--dry-run", action="store_true")

    categories_parser = subparsers.add_parser("categories")
    categories_parser.add_argument("--dry-run", action="store_true")

    case_parser = subparsers.add_parser("case")
    case_parser.add_argument("--case-id", required=True)
    case_parser.add_argument("--dry-run", action="store_true")

    templates_parser = subparsers.add_parser("templates")
    templates_parser.add_argument("--dry-run", action="store_true")

    args = parser.parse_args()
    base_url = clean_base_url(args.base_url)
    if args.command == "search":
        query = {"q": args.query, "limit": args.limit}
        if args.offset:
            query["offset"] = args.offset
        if args.category:
            query["category"] = args.category
        print_or_dry_run(f"{base_url}/inspiration/search?{urllib.parse.urlencode(query)}", args.dry_run, normalize_search)
    elif args.command == "categories":
        print_or_dry_run(f"{base_url}/inspiration/categories", args.dry_run)
    elif args.command == "case":
        print_or_dry_run(f"{base_url}/inspiration/cases/{urllib.parse.quote(args.case_id)}", args.dry_run)
    elif args.command == "templates":
        print_or_dry_run(f"{base_url}/free-image/api/templates", args.dry_run, normalize_templates)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
